File: src/main/java/org/github/erikzielke/gotoproject/GoToProjectApplicationComponent.java
package org.github.erikzielke.gotoproject;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.components.BaseComponent;
import com.intellij.openapi.components.PersistentStateComponent;
import com.intellij.openapi.components.State;
import com.intellij.openapi.components.Storage;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.wm.WindowManager;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Created by Erik on 10-02-14.
 */
@State(name = "GoToProject", storages = {@Storage("$APP_CONFIG$/GoToProjectWindow.xml")})
public class GoToProjectApplicationComponent implements PersistentStateComponent<GoToProjectWindowSettings>, BaseComponent {

    private GoToProjectWindowSettings state;
    private Project focusedBefore;

    public void initComponent() {
        if (state == null) {
            state = new GoToProjectWindowSettings();
        }
        if (state.isIncludeRecent()) {
            WindowManager.getInstance().findVisibleFrame();
        }

    }

    static GoToProjectApplicationComponent getInstance() {
        return ApplicationManager.getApplication()
                                 .getComponent(GoToProjectApplicationComponent.class);
    }

    @NotNull
    public String getComponentName() {
        return "Go To Project";
    }


    Project getFocusedBefore() {
        return focusedBefore;
    }

    @Nullable
    @Override
    public GoToProjectWindowSettings getState() {
        return state;
    }

    @Override
    public void loadState(@NotNull GoToProjectWindowSettings goToProjectWindowSettings) {
        this.state = goToProjectWindowSettings;
    }

    void setFocusedBefore(Project focusedBefore) {
        this.focusedBefore = focusedBefore;
    }
}

File: src/main/kotlin/org/github/erikzielke/gotoproject/GoToProjectApplicationComponent.kt
package org.github.erikzielke.gotoproject

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.components.PersistentStateComponent
import com.intellij.openapi.components.Service
import com.intellij.openapi.components.State
import com.intellij.openapi.components.Storage
import com.intellij.openapi.project.Project

@Service
@State(name = "GoToProject", storages = [Storage(value = "GoToProjectWindow.xml")])
class GoToProjectApplicationComponent : PersistentStateComponent<GoToProjectWindowSettings?> {
    private var state: GoToProjectWindowSettings = GoToProjectWindowSettings()
    var focusedBefore: Project? = null
    var lastFocusLost: Project? = null

    override fun getState(): GoToProjectWindowSettings = state

    override fun loadState(state: GoToProjectWindowSettings) {
        this.state = state
    }

    companion object {
        @JvmStatic
        val instance: GoToProjectApplicationComponent
            get() = ApplicationManager.getApplication().getService(GoToProjectApplicationComponent::class.java)
    }
}
